### PR TITLE
[SPARK-38112][SQL] Use error classes in the execution errors of date/timestamp handling

### PR DESCRIPTION
**Pull request by ivoson**
### What changes were proposed in this pull request?
Migrate the following errors in QueryExecutionErrors onto use error classes:
- sparkUpgradeInReadingDatesError => INCONSISTENT_BEHAVIOR_CROSS_VERSION
- sparkUpgradeInWritingDatesError => INCONSISTENT_BEHAVIOR_CROSS_VERSION
- timeZoneIdNotSpecifiedForTimestampTypeError => UNSUPPORTED_OPERATION
- cannotConvertOrcTimestampToTimestampNTZError => UNSUPPORTED_OPERATION

### Why are the changes needed?
Porting date/timestamp execute errors to new error framework.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @MaxGekk FYI

File: core/src/main/scala/org/apache/spark/SparkException.scala [hunk 1]
```diff
@@ -71,9 +71,35 @@ private[spark] case class ExecutorDeadException(message: String)
...
+    message: String,
+    cause: Throwable,
+    errorClass: Option[String],
+    messageParameters: Array[String])
+  extends RuntimeException(message, cause) with SparkThrowable {
+
+  def this(version: String, message: String, cause: Throwable) =
+    this (
+      version = version,
+      message = s"You may get a different result due to the upgrading of Spark $version: $message",
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/SparkException.scala:85]: Let's avoid error message duplication. error-classes.json should be one source of truth:
```
  "INCONSISTENT_BEHAVIOR_CROSS_VERSION" : {
    "message" : [ "You may get a different result due to the upgrading of Spark %s: %s" ]
  },
```

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 2]
```diff
@@ -171,4 +181,72 @@ class QueryExecutionErrorsSuite extends QueryTest with SharedSparkSession {
...
+    withSQLConf(SQLConf.PARQUET_REBASE_MODE_IN_READ.key -> EXCEPTION.toString) {
+      val fileName = "before_1582_date_v2_4_5.snappy.parquet"
+      val filePath = getResourceParquetFilePath("test-data/" + fileName)
+      val e = intercept[SparkException] {
+        spark.read.parquet(filePath).collect()
+      }.getCause.asInstanceOf[SparkUpgradeException]
+
+      assert(e.getErrorClass === "INCONSISTENT_BEHAVIOR_CROSS_VERSION")
+      assert(e.getMessage
+        .startsWith("You may get a different result due to the upgrading of Spark 3.0: \n" +
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:198]: The message confuses slightly. We run Spark 3.3.0-SNAPSHOT (almost 3.3) and try to read a file written by Spark 2.4 but the message says: due to the **upgrading of Spark 3.0**.

Should be somehow: ... due to upgrading to Spark >= 3.0?

Comment (ivoson) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:198]: Spark >= 3.0 sounds good. Will address in next iteration.

File: core/src/main/scala/org/apache/spark/SparkException.scala (hunk 1 above)
Comment (ivoson) [reply core/src/main/scala/org/apache/spark/SparkException.scala:85]: Thanks @MaxGekk. Will fix this.

**Comment by ivoson** [comment]
Hey @MaxGekk . Comments addressed, please take a look when you have time. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 3]
```diff
@@ -181,4 +191,72 @@ class QueryExecutionErrorsSuite extends QueryTest with SharedSparkSession {
...
+    withSQLConf(SQLConf.PARQUET_REBASE_MODE_IN_READ.key -> EXCEPTION.toString) {
+      val fileName = "before_1582_date_v2_4_5.snappy.parquet"
+      val filePath = getResourceParquetFilePath("test-data/" + fileName)
+      val e = intercept[SparkException] {
+        spark.read.parquet(filePath).collect()
+      }.getCause.asInstanceOf[SparkUpgradeException]
+
+      assert(e.getErrorClass === "INCONSISTENT_BEHAVIOR_CROSS_VERSION")
+      assert(e.getMessage
+        .startsWith("You may get a different result due to the upgrading to Spark >= 3.0: \n" +
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:208]: Could you check entire error message, please. See the motivation for that in PR's description  https://github.com/apache/spark/pull/35416

**Review by MaxGekk — COMMENTED** [review]
LGTM except of a comment.

**Comment by MaxGekk** [comment]
@ivoson Could you fix coding style:
```
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:215: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:237: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:238: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:239: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:240: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:241: File line length exceeds 100 characters
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:242: File line length exceeds 100 characters
```

**Comment by ivoson** [comment]
> @ivoson Could you fix coding style:
> 
> ```
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:215: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:237: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:238: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:239: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:240: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:241: File line length exceeds 100 characters
> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:242: File line length exceeds 100 characters
> ```

Thanks @MaxGekk . Fixed.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @ivoson.

**Comment by ivoson** [comment]
Thanks for the review. @MaxGekk

**Pull request closed without merging**